src/common/commonutils/PackageUtils.c

```diff
@@ -90,51 +108,173 @@ static int CheckOrInstallPackage(const char* commandTemplate, const char* packag
+        status = CheckAllPackages(commandTmeplateZypper, g_zypper, &results, log);
     }
 
-    if (0 == status)
+    if ((0 == status) && (NULL != results))
     {
-        OsConfigLogInfo(log, "IsPackageInstalled: package '%s' is installed", packageName);
+        if (NULL != (buffer = DuplicateString(results)))
+        {
+            FREE_MEMORY(g_installedPackagesCache);
+            g_installedPackagesCache = buffer;
+        }
+        else
+        {
+            // Leave the cache as-is, just log the error
+            OsConfigLogError(log, "UpdateInstalledPackagesCache: out of memory");
+            status = ENOMEM;
+        }
     }
     else
     {
-        OsConfigLogInfo(log, "IsPackageInstalled: package '%s' is not found (%d, errno: %d)", packageName, status, errno);
+        // Leave the cache as-is, we can still use it even if it's stale
+        status = status ? status : ENOENT;
+        OsConfigLogInfo(log, "UpdateInstalledPackagesCache: enumerating all packages failed with %d", status);
+    }
+
+    FREE_MEMORY(results);
+
+    return status;
+}
+
+int IsPackageInstalled(const char* packageName, OsConfigLogHandle log)
+{
+    const char* searchTemplateDpkgRpm = "\n%s\n";
+    const char* searchTemplateYumDnf = "\n%s.x86_64\n";
+    const char* searchTemplateZypper = "| %s ";
+
+    char* searchTarget = NULL;
+    int status = 0;
+
+    if ((NULL == packageName) || (0 == strlen(packageName)))
+    {
+        OsConfigLogError(log, "IsPackageInstalled called with an invalid argument");
+        return EINVAL;
+    }
+
+    CheckPackageManagersPresence(log);
+
+    if ((0 != g_updateInstalledPackagesCache) || (NULL == g_installedPackagesCache))
+    {
+        g_updateInstalledPackagesCache = 0;
+
+        if (0 != (status = UpdateInstalledPackagesCache(log)))
+        {
+            OsConfigLogInfo(log, "IsPackageInstalled(%s) failed (UpdateInstalledPackagesCache failed)", packageName);
+        }
+    }
+
+    if (NULL == g_installedPackagesCache)
+    {
+        OsConfigLogError(log, "IsPackageInstalled: cannot check for '%s' presence without cache", packageName);
+        status = ENOENT;
+    }
+    else if (0 == status)
+    {
+        if (g_aptGetIsPresent || g_dpkgIsPresent || g_rpmIsPresent)
+        {
+            searchTarget = FormatAllocateString(searchTemplateDpkgRpm, packageName);
+        }
+        else if (g_tdnfIsPresent || g_dnfIsPresent || g_yumIsPresent)
+        {
+            searchTarget = FormatAllocateString(searchTemplateYumDnf, packageName);
+        }
+        else //if (g_zypperIsPresent)
+        {
+            searchTarget = FormatAllocateString(searchTemplateZypper, packageName);
+        }
+
+        if (NULL == searchTarget)
+        {
+            OsConfigLogError(log, "IsPackageInstalled: out of memory");
+            status = ENOMEM;
+        }
+        else
+        {
+            if (NULL != strstr(g_installedPackagesCache, searchTarget))
```

Comment:
We should make sure the `searchTarget` is the whole word search as you may be picking a partial name of a package? eg.
 * `python`, `python3`
 * `ninja`, `ninja-build`

Reply:
This is taken care of the "\n...\n" for the dpkg and rpm and other ways for the other managers. It's tested by our CI tests actually. We only get exact name packages. If we look for "bluetooth" it will be just that, no "bluetooth-nx" etc.